File: src/main/java/com/vc/medium/L249.scala
package com.vc.medium

object L249 {
  def groupStrings(strings: Array[String]): List[List[String]] = {
    import scala.collection.mutable
    val map = new mutable.HashMap[String, mutable.ListBuffer[String]]()
    val res = new mutable.ListBuffer[List[String]]()
    strings.foreach(s => {
      val offset = s(0) - 'a'
      var sum = ""
      s.foreach(ch => {
        var c = ch - offset
        if(c < 'a') c += 26
        sum += c
      })
      val list = map.getOrElse(sum, new mutable.ListBuffer[String]())
      list += s
      map.put(sum, list)
    })
    map.foreach(x => {
      res += x._2.toList.sorted
    })
    res.toList
  }
}

File: src/main/java/com/vc/medium/L249.java
package com.vc.medium;

import java.util.*;

class L249 {
    public List<List<String>> groupStrings(String[] strings) {
        Map<String, List<String>> map = new HashMap<> ();
        for (String str : strings) {
            String key = getKey(str);
            map.putIfAbsent(key, new ArrayList<>());
            map.get(key).add(str);
        }
        return new ArrayList<List<String>> (map.values());
    }

    private String getKey(String word) {
        StringBuilder sb = new StringBuilder();
        if (word.isEmpty()) return sb.toString();
        char first = word.charAt(0);
        for (int i = 0; i < word.length(); i++) {
            int ch = word.charAt(i) - first;
            if(ch < 0) ch += 26;
            sb.append((char)ch);
        }
        return sb.toString();
    }
}
